### PR TITLE
Update TrajectoryPrediction.java

### PR DESCRIPTION
**Pull request by ManavToor**
Adding better drag calculations by looking up Ansys results

- - -
This change is [Reviewable](https://reviewable.io/reviews/waterloo-rocketry/or-airbrake-plugin/2)

### CODE REVIEW

File: src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java [hunk 1]
```diff
@@ -59,17 +61,60 @@ private static double air_density(double altitude) {
...
+    }
 
-        // Fd = 0.5 * Cd * A * p * v^2
-        return -0.5 * drag_coef * ref_area * fluid_dens * velocity * velocity;
+    /**@return drag force value based off Ansys simulations
+     * @param velocity is used to lookup drag value
+     * @param extension MUST BE EITHER 0, 50, 100. USed to determing which function to use and to adjust for rocket area
+     * air_density is adjusted for since all sims were calculated at 1000m
+     * */
+    private static double lookup_drag(int extension, double velocity) {
```
Comment (Joe-Joe-Joe-Joe) [inline src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java]: We should use a different parameter name for "fixed extensions for looking up" and "continuum of commanded extensions". Or I guess in this case `int extension` vs `double extension`

File: src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java [hunk 2]
```diff
@@ -59,17 +61,60 @@ private static double air_density(double altitude) {
...
+     * @param altitude used to adjust fluid density since Ansys sims were calculated at 1000m
+     * */
+    private static double interpolate_drag(double extension, double velocity, double altitude) {
+        if (extension == 100 || extension == 50 ||  extension == 0) {
+            int ext = (int)extension;
+            return air_density(altitude) * rocket_area(ext) * lookup_drag(ext, velocity);
+        }
+        else if (extension > 50.0) {
+            double diff = lookup_drag(100, velocity) - lookup_drag(50, velocity);
+            extension = (extension - 50) * 2;
```
Comment (Joe-Joe-Joe-Joe) [inline src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java]: Its too late for me to do math in my head - why the floating factor of 2?

File: src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java [hunk 3]
```diff
@@ -37,7 +37,9 @@ private static double gravitational_acceleration(double altitude) {
         return 9.80665 * Math.pow(6371009 / ( 6371009 + altitude), 2);
     }
 
-    /** @return air density from altitude */
+    /** @return air density (kg/m^3)
+     * @param altitude (m)
+     * Same method used for PyAnsys simulations, verified to be correct*/
```
Comment (Joe-Joe-Joe-Joe) [inline src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java]: Can we please move some of these magic numbers to constants?

**Review by Joe-Joe-Joe-Joe — CHANGES_REQUESTED** [review]
Excellent work, I'm probably going to start working with this code tomorrow. I would like some code cleanup before merging PR, since I don't know that the code will be touched much after this.

**Review by patrick-gu — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/or-airbrake-plugin/2)* status: all files reviewed, 8 unresolved discussions (waiting on @celery6 and @ManavToor)

___
*[`src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java` line 76 at r1](https://reviewable.io/reviews/waterloo-rocketry/or-airbrake-plugin/2#-NrxASnrAa7K_C_b1qgE:-NrxASnrAa7K_C_b1qgF:b-of7ij0) ([raw file](https://github.com/waterloo-rocketry/or-airbrake-plugin/blob/7b1063e7c1a8456e479d6399479c593f5696bbce/src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java#L76)):*
> ```Java
>      * @param velocity is used to lookup drag value
>      * @param extension MUST BE EITHER 0, 50, 100. USed to determing which function to use and to adjust for rocket area
>      * air_density is adjusted for since all sims were calculated at 1000m
> ```

does this mean that drag is linear with air density?
___
*[`src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java` line 84 at r1](https://reviewable.io/reviews/waterloo-rocketry/or-airbrake-plugin/2#-NrxArKMDB1TpBbVWbQe:-NrxArKMDB1TpBbVWbQf:bj7pgag) ([raw file](https://github.com/waterloo-rocketry/or-airbrake-plugin/blob/7b1063e7c1a8456e479d6399479c593f5696bbce/src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java#L84)):*
> ```Java
>         } else if (extension == 50) {
>             drag = (0.0045 * velocity * velocity) + (0.1031 * velocity) - 3.8231;
>         } else { // extension == 100
> ```

maybe throw an IllegalArgumentException here
___
*[`src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java` line 85 at r1](https://reviewable.io/reviews/waterloo-rocketry/or-airbrake-plugin/2#-NrxABHA4tro32bNE79_:-NrxABHA4tro32bNE79a:b6dmt8v) ([raw file](https://github.com/waterloo-rocketry/or-airbrake-plugin/blob/7b1063e7c1a8456e479d6399479c593f5696bbce/src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java#L85)):*
> ```Java
>             drag = (0.0045 * velocity * velocity) + (0.1031 * velocity) - 3.8231;
>         } else { // extension == 100
>             drag = (0.006*velocity*velocity) + (0.1038*velocity) - 4.2522;
> ```

later on we can separate this out to an array, I think it's fine for now
___
*[`src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java` line 91 at r1](https://reviewable.io/reviews/waterloo-rocketry/or-airbrake-plugin/2#-NrxAig_6u4jlr7Fd4Zv:-NrxAig_6u4jlr7Fd4Zw:b-rdirgd) ([raw file](https://github.com/waterloo-rocketry/or-airbrake-plugin/blob/7b1063e7c1a8456e479d6399479c593f5696bbce/src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java#L91)):*
> ```Java
> 
>     /**@return drag force acting on rocket
>      * @param extension of air brakes, used for adjusting rocket area and iterpolating Ansys sims
> ```

do you think extension should be 0 to 100 or 0 to 1? I feel like 0 to 1 would be more simple
___
*[`src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java` line 105 at r1](https://reviewable.io/reviews/waterloo-rocketry/or-airbrake-plugin/2#-NrxB4ue-92YpitUpr4s:-NrxB4ue-92YpitUpr4t:b-anxaol) ([raw file](https://github.com/waterloo-rocketry/or-airbrake-plugin/blob/7b1063e7c1a8456e479d6399479c593f5696bbce/src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java#L105)):*
> ```Java
>             return air_density(altitude) * rocket_area(extension) * (diff * extension) + lookup_drag(50, velocity);
>         }
>         else { // extension < 50
> ```

do you plan on rewriting this when we keep adding more data? you might want to have a look at my linear interpolation calculator, which would make this more concise (though you may want to change it, not sure if it applies exactly to this situation)

**Comment by Joe-Joe-Joe-Joe** [comment]
> does this mean that drag is linear with air density?

it is approximately linear within our flight regime

File: src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java (hunk 2 above)
Comment (ManavToor) [reply src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java]: Maps extension 50-100% to 0-100% to estimate y-value between the 50% and 100% curves

ie. 75% airbrake extension is halfway (50%) between the 50 and 100 curves

**Review by celery6 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/or-airbrake-plugin/2)* status: 0 of 6 files reviewed, 9 unresolved discussions (waiting on @Joe-Joe-Joe-Joe, @ManavToor, and @patrick-gu)

___
*[`src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java` line 146 at r3](https://reviewable.io/reviews/waterloo-rocketry/or-airbrake-plugin/2#-NsEOiym7cr0RE3N5QrC:-NsEOiym7cr0RE3N5QrD:b-fbfsmb) ([raw file](https://github.com/waterloo-rocketry/or-airbrake-plugin/blob/6222545e1d00752f007b958f941310bcd884cb11/src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java#L146)):*
> ```Java
>      * @return Cd value corresponding to interpolated drag force
>      */
>     public static double interpolate_cd(double extension, double velocity, double altitude){
> ```

We should create a separate function like convertDragToCd(). Then the interpolate_drag() can be reused by airbrakes, and the interpolation implementation can change without affecting Cd calculation.

**Review by patrick-gu — CHANGES_REQUESTED** [review]
> Reviewed 5 of 6 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/or-airbrake-plugin/2)* status: all files reviewed, 13 unresolved discussions (waiting on @Joe-Joe-Joe-Joe and @ManavToor)

___
*[`src/main/java/com/waterloorocketry/airbrakeplugin/AirbrakePluginSimulationListener.java` line 78 at r3](https://reviewable.io/reviews/waterloo-rocketry/or-airbrake-plugin/2#-NsODpO66tSikFQGjseK:-NsODpO66tSikFQGjseL:bb6kw3a) ([raw file](https://github.com/waterloo-rocketry/or-airbrake-plugin/blob/6222545e1d00752f007b958f941310bcd884cb11/src/main/java/com/waterloorocketry/airbrakeplugin/AirbrakePluginSimulationListener.java#L78)):*
> ```Java
> 		
> 		// Override CD only during coast, and until velocity is too small for the drag tabulation to be accurate
> 		if (status.isMotorIgnited() && !status.isApogeeReached() && velocity > 23.5) {
> ```

I think this overrides cd even before burnout
___
*[`src/main/java/com/waterloorocketry/airbrakeplugin/controller/PIDController.java` line 29 at r3](https://reviewable.io/reviews/waterloo-rocketry/or-airbrake-plugin/2#-NsOE7Ht9-K3oq2lmzSH:-NsOE7Ht9-K3oq2lmzSI:b-xdov2v) ([raw file](https://github.com/waterloo-rocketry/or-airbrake-plugin/blob/6222545e1d00752f007b958f941310bcd884cb11/src/main/java/com/waterloorocketry/airbrakeplugin/controller/PIDController.java#L29)):*
> ```Java
>     @Override
>     public double calculateTargetExt(double[] flightData, double time, double extension) {
>         double altitude = TrajectoryPrediction.get_max_altitude(flightData[5], flightData[2], extension, ROCKET_BURNOUT_MASS); //z displacement, z velocity, ...; +Z is "up" in OR
> ```

I created an array of data as Joe requested, but if it gets too annoying, it can probably be just an object for convenience (so that we don't have to keep looking up indices as magic numbers)
___
*[`src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java` line 119 at r3](https://reviewable.io/reviews/waterloo-rocketry/or-airbrake-plugin/2#-NsOEggU54r2v4P6msNv:-NsOEggU54r2v4P6msNw:bda777z) ([raw file](https://github.com/waterloo-rocketry/or-airbrake-plugin/blob/6222545e1d00752f007b958f941310bcd884cb11/src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java#L119)):*
> ```Java
>             extension *= 2;
>             drag = air_density(altitude) * rocket_area(extension) * (diff * extension) + lookup_drag(0, velocity);
>         } */
> ```

maybe remove this commented out code later?
___
*[`src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java` line 176 at r3](https://reviewable.io/reviews/waterloo-rocketry/or-airbrake-plugin/2#-NsOEuzi2kLEzqhFB45y:-NsOEuzi2kLEzqhFB45z:b-b414lk) ([raw file](https://github.com/waterloo-rocketry/or-airbrake-plugin/blob/6222545e1d00752f007b958f941310bcd884cb11/src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java#L176)):*
> ```Java
>         double h = 0.05; // interval of change for rk4
>         double prevAlt = 0.0; // variable to store previous altitude
>         double[] states = {altitude, velocity}; // array to store altitude and velocity
> ```

nit, but not sure if storing an array is the cleanest way to do it - maybe an object would be cleaner?

**Review by celery6 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/or-airbrake-plugin/2)* status: all files reviewed, 13 unresolved discussions (waiting on @Joe-Joe-Joe-Joe, @ManavToor, and @patrick-gu)

___
*[`src/main/java/com/waterloorocketry/airbrakeplugin/AirbrakePluginSimulationListener.java` line 78 at r3](https://reviewable.io/reviews/waterloo-rocketry/or-airbrake-plugin/2#-NsODpO66tSikFQGjseK:-NsOlq1u2gRhNo3wtNV1:b-eivk6z) ([raw file](https://github.com/waterloo-rocketry/or-airbrake-plugin/blob/6222545e1d00752f007b958f941310bcd884cb11/src/main/java/com/waterloorocketry/airbrakeplugin/AirbrakePluginSimulationListener.java#L78)):*
<details><summary><i>Previously, patrick-gu (Patrick Gu) wrote…</i></summary><blockquote>

I think this overrides cd even before burnout
</blockquote></details>

Fuck I meant to delete that before committing. Pushed fix.

**Review by patrick-gu — CHANGES_REQUESTED** [review]
> Reviewed 4 of 4 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/or-airbrake-plugin/2)* status: all files reviewed, 15 unresolved discussions (waiting on @Joe-Joe-Joe-Joe and @ManavToor)

___
*[`src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java` line 7 at r5](https://reviewable.io/reviews/waterloo-rocketry/or-airbrake-plugin/2#-NtNxuEE280L6T6R5w1A:-NtNxuEE280L6T6R5w1B:b-hg13jn) ([raw file](https://github.com/waterloo-rocketry/or-airbrake-plugin/blob/f45fe96d989255f34e9ac9b1c4416a9c2919877a/src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java#L7)):*
> ```Java
>     private static final double EARTH_MEAN_RADIUS = 6371009;
>     private static final double AIRBRAKES_MAX_AREA = 0.004993538; //m^2
>     private static final double ROCKET_BASE_AREA = 0.0182412538; //in^2
> ```

this is in^2?
___
*[`src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java` line 9 at r5](https://reviewable.io/reviews/waterloo-rocketry/or-airbrake-plugin/2#-NtNyENXD7cmSHuJ8RNw:-NtNyENXD7cmSHuJ8RNx:b-tq2dv) ([raw file](https://github.com/waterloo-rocketry/or-airbrake-plugin/blob/f45fe96d989255f34e9ac9b1c4416a9c2919877a/src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java#L9)):*
> ```Java
>     private static final double ROCKET_BASE_AREA = 0.0182412538; //in^2
>     private static final double SIM_ALTITUDE = 1000; //All drag sims conducted at 1000m above sea level
>     private static final double TOL = 0.00001;
> ```

can we maybe add a doc comment about this
___
*[`src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java` line 110 at r5](https://reviewable.io/reviews/waterloo-rocketry/or-airbrake-plugin/2#-NtNyM1aAsvWApvlCZJn:-NtNyM1aAsvWApvlCZJo:b-j18z4o) ([raw file](https://github.com/waterloo-rocketry/or-airbrake-plugin/blob/f45fe96d989255f34e9ac9b1c4416a9c2919877a/src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java#L110)):*
> ```Java
>         double x_1;
> 
>         if(extension < TOL || extension > 1 + TOL){
> ```

if we're doing this then shouldn't everything be offset?

**Comment by Joe-Joe-Joe-Joe** [comment]
> if we're doing this then shouldn't everything be offset?

The reason I did that was for cases where a floating point error could cause undefined behaviour. Hope this will be addressed in the interpolation rewrite. Unfortunately representing extension amt as a float and then doing comparison necessitates this sort of thing. 

> can we maybe add a doc comment about this

Sure. Can you do that? You seem to know what you want.

> this is in^2?

Nope. Fixed.

**Comment by Joe-Joe-Joe-Joe** [comment]
Re: discussions on changing arrays to objects - I 100% agree. I assume that is fairly trivial for one of you to do?

**Review by Joe-Joe-Joe-Joe — APPROVED** [review]
On my side I'm happy with the code - it does the job as expected.

**Review by celery6 — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/or-airbrake-plugin/2#-:-NtTdIcHBq4kJvq4amIb:bnfp4nl)* status: 1 of 6 files reviewed, 14 unresolved discussions (waiting on @Joe-Joe-Joe-Joe, @ManavToor, and @patrick-gu)

___
*[`src/main/java/com/waterloorocketry/airbrakeplugin/controller/PIDController.java` line 29 at r3](https://reviewable.io/reviews/waterloo-rocketry/or-airbrake-plugin/2#-NsOE7Ht9-K3oq2lmzSH:-NtTb-e95W2pK4x4caZF:bg1g6im) ([raw file](https://github.com/waterloo-rocketry/or-airbrake-plugin/blob/6222545e1d00752f007b958f941310bcd884cb11/src/main/java/com/waterloorocketry/airbrakeplugin/controller/PIDController.java#L29)):*
<details><summary><i>Previously, patrick-gu (Patrick Gu) wrote…</i></summary><blockquote>

I created an array of data as Joe requested, but if it gets too annoying, it can probably be just an object for convenience (so that we don't have to keep looking up indices as magic numbers)
</blockquote></details>

updated to use a custom data holder class instead of array.
___
*[`src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java` line 9 at r5](https://reviewable.io/reviews/waterloo-rocketry/or-airbrake-plugin/2#-NtNyENXD7cmSHuJ8RNw:-NtTdC7FCFkx0HNOxxUR:b-dzqdsd) ([raw file](https://github.com/waterloo-rocketry/or-airbrake-plugin/blob/f45fe96d989255f34e9ac9b1c4416a9c2919877a/src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java#L9)):*
<details><summary><i>Previously, patrick-gu (Patrick Gu) wrote…</i></summary><blockquote>

can we maybe add a doc comment about this
</blockquote></details>

added brief comment; this is fairly standard floating point procedure so not too much explanation

**Review by patrick-gu — APPROVED** [review]
:lgtm:

> Reviewed 4 of 5 files at r7, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/or-airbrake-plugin/2#-:-NtWrYwr0gj6SFzUlu1U:bnfp4nl)* status: all files reviewed, 5 unresolved discussions (waiting on @Joe-Joe-Joe-Joe and @ManavToor)

___
*[`src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java` line 105 at r1](https://reviewable.io/reviews/waterloo-rocketry/or-airbrake-plugin/2#-NrxB4ue-92YpitUpr4s:-NtWrG3Y3FiVEmkEsLQX:bhpa9v8) ([raw file](https://github.com/waterloo-rocketry/or-airbrake-plugin/blob/7b1063e7c1a8456e479d6399479c593f5696bbce/src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java#L105)):*
<details><summary><i>Previously, patrick-gu (Patrick Gu) wrote…</i></summary><blockquote>

do you plan on rewriting this when we keep adding more data? you might want to have a look at my linear interpolation calculator, which would make this more concise (though you may want to change it, not sure if it applies exactly to this situation)
</blockquote></details>

we are going to change this later
___
*[`src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java` line 9 at r5](https://reviewable.io/reviews/waterloo-rocketry/or-airbrake-plugin/2#-NtNyENXD7cmSHuJ8RNw:-NtWqX1z9lNB_Mw_D0Lr:b-etoo6q) ([raw file](https://github.com/waterloo-rocketry/or-airbrake-plugin/blob/f45fe96d989255f34e9ac9b1c4416a9c2919877a/src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java#L9)):*
<details><summary><i>Previously, celery6 (celery) wrote…</i></summary><blockquote>

added brief comment; this is fairly standard floating point procedure so not too much explanation
</blockquote></details>

this probably won't be too necessary later on since we probably will not need the equality comparison - just inequalities
___
*[`src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java` line 110 at r5](https://reviewable.io/reviews/waterloo-rocketry/or-airbrake-plugin/2#-NtNyM1aAsvWApvlCZJn:-NtWrKE-7sOPzWgP7i42:b269i8t) ([raw file](https://github.com/waterloo-rocketry/or-airbrake-plugin/blob/f45fe96d989255f34e9ac9b1c4416a9c2919877a/src/main/java/com/waterloorocketry/airbrakeplugin/controller/TrajectoryPrediction.java#L110)):*
<details><summary><i>Previously, patrick-gu (Patrick Gu) wrote…</i></summary><blockquote>

if we're doing this then shouldn't everything be offset?
</blockquote></details>

we can change this later with more linear interpolation

**Pull request merged**